include/deal.II/base/symmetric_tensor.h

```diff
@@ -2753,6 +2781,99 @@ invert (const SymmetricTensor<4,3,double> &t);
 
 
 /**
+ * Full contraction of three tensors: Return a scalar number that is the
+ * result of a full contraction of a symmetric tensor @p left of rank @p rank_1,
+ *  a tensor @p middle of rank $(\text{rank}_1+\text{rank}_2)$ and a tensor @p
+ * right of rank @p rank_2:
+ * @f[
+ *   \sum_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{left}_{i_1,\ldots,i_{r1}}
+ *   \text{middle}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{right}_{j_1,\ldots,j_{r2}}
+ * @f]
+ *
+ * @relates Tensor
+ * @relates SymmetricTensor
+ * @author Jean-Paul Pelteret, Matthias Maier, 2017
+ */
+template <int rank_1, int rank_2, int dim,
+          typename T1, typename T2, typename T3>
+typename ProductType<T1, typename ProductType<T2, T3>::type>::type
+contract3 (const SymmetricTensor<rank_1, dim, T1> &left,
+           const Tensor<rank_1 + rank_2, dim, T2> &middle,
+           const Tensor<rank_2, dim, T3>          &right)
```

Comment:
What about fully "templifying" the contract3 function?
The problem at the moment is that with every tensor version and every combination you have to do an additional contract3 function definition.

It should be enough to simply replace the `contract3` function in `tensor.h` with the following:
```
1664 template <template<int, int, typename> class TensorT1,                           
1665           template<int, int, typename> class TensorT2,                           
1666           template<int, int, typename> class TensorT3,                           
1667           int rank1, int rank2, int dim,                                         
1668           typename T1, typename T2, typename T3>                                 
1669 typename ProductType<T1, typename ProductType<T2, T3>::type>::type               
1670 contract3 (const TensorT1<rank1, dim, T1> &left,                                 
1671            const TensorT2<rank1 + rank2, dim, T2> &middle,                       
1672            const TensorT3<rank2, dim, T3> &right)                                
1673 {                                                                                
1674   typedef typename ProductType<T1, typename ProductType<T2, T3>::type>::type     
1675   return_type;                                                                   
1676   return TensorAccessors::contract3<rank1, rank2, dim, return_type>(             
1677            left, middle, right);
```